mobile/cafes_test.go

```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// register with cafe
+	url := "http://127.0.0.1:" + cafesTestVars.cafeApiPort
+	err = cafesTestVars.mobile.RegisterCafe(url, token)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// add some data
+	err = addTestData(cafesTestVars.mobile)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// manually flush until empty
+	err = flush(10)
+	if err != nil {
+		fmt.Println(err.Error())
+	}
+}
+
+func addTestData(m *Mobile) error {
+	thrd, err := addTestThread(m, &pb.AddThreadConfig{
+		Key:  ksuid.New().String(),
+		Name: "test",
+		Schema: &pb.AddThreadConfig_Schema{
+			Preset: pb.AddThreadConfig_Schema_MEDIA,
+		},
+		Type:    pb.Thread_PRIVATE,
+		Sharing: pb.Thread_INVITE_ONLY,
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = m.addFiles([]string{"../mill/testdata/image.jpeg"}, thrd.Id, "hi")
+	if err != nil {
+		return err
+	}
+
+	_, err = m.AddMessage(thrd.Id, "hi")
+	if err != nil {
+		return err
+	}
+
+	hash, err := m.addFiles([]string{"../mill/testdata/image.png"}, thrd.Id, "hi")
+	if err != nil {
+		return err
+	}
+
+	_, err = m.AddComment(hash.B58String(), "nice")
+	if err != nil {
+		return err
+	}
+
+	hash, err = m.addFiles([]string{"../mill/testdata/image.jpeg", "../mill/testdata/image.png"}, thrd.Id, "hi")
+	if err != nil {
+		return err
+	}
+
+	_, err = m.AddLike(hash.B58String())
+	if err != nil {
+		return err
+	}
+
+	_, err = m.AddMessage(thrd.Id, "bye")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+/*
+Handle the request queue.
+  1. List some requests
+  2. Write the HTTP request for each
+  3. Handle them (set to pending, send to cafe)
+  4. Delete failed (reties not handled here)
+  5. Set successful to complete
+*/
+func flushCafeRequests(limit int) (int, error) {
+	var count int
+	res, err := cafesTestVars.mobile.CafeRequests(limit)
+	if err != nil {
+		return count, err
+	}
+	groups := new(pb.Strings)
+	err = proto.Unmarshal(res, groups)
+	if err != nil {
+		return count, err
+	}
+	count = len(groups.Values)
+
+	// write the req for each group
+	reqs := make(map[string]*pb.CafeHTTPRequest)
+	for _, g := range groups.Values {
+		res, err = cafesTestVars.mobile.WriteCafeRequest(g)
+		if err != nil {
+			return count, err
+		}
+		req := new(pb.CafeHTTPRequest)
+		err = proto.Unmarshal(res, req)
+		if err != nil {
+			return count, err
+		}
+		reqs[g] = req
+	}
+
+	// mark each as pending (new loops for clarity)
+	for g := range reqs {
+		err = cafesTestVars.mobile.CafeRequestPending(g)
+		if err != nil {
+			return count, err
+		}
+	}
+
+	// handle each
+	for g, req := range reqs {
+		res, err := handleReq(req)
+		if err != nil {
+			return count, err
+		}
+		if res.StatusCode >= 300 {
+			fmt.Printf("got bad status: %d\n", res.StatusCode)
+			err = cafesTestVars.mobile.FailCafeRequest(g)
+		} else {
+			err = cafesTestVars.mobile.CompleteCafeRequest(g)
+		}
+		if err != nil {
+			return count, err
+		}
+		res.Body.Close()
+	}
+	return count, nil
+}
+
+func flush(batchSize int) error {
```

Comment:
This test method here just manually recursively handles all the requests after adding a bunch of data.